### PR TITLE
Implement registration client to read old version's metadata

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/NuGetGallery/issues/6446

Also fix two bugs in the catalog client.

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/Registration/IRegistrationClient.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace NuGet.Protocol.Registration
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/Registration/IRegistrationClient.cs]: >NuGet.Protocol.Registration [](start = 10, length = 27)

What do you think of renaming `NuGet.Protocol.Catalog` to `NuGet.Protocol` and moving the registration libraries there?

File: src/NuGet.Services.AzureSearch/Registration/RegistrationClient.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        public RegistrationClient(ISimpleHttpClient simpleHttpClient)
+        {
+            _simpleHttpClient = simpleHttpClient ?? throw new ArgumentNullException(nameof(simpleHttpClient));
+        }
+
+        public async Task<RegistrationIndex> GetIndexAsync(string indexUrl)
+        {
+            var result = await _simpleHttpClient.DeserializeUrlAsync<RegistrationIndex>(indexUrl);
+            if (result.StatusCode == HttpStatusCode.NotFound)
+            {
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/Registration/RegistrationClient.cs]: It's not clear this returns null if the index doesn't exist. I would either:

1. Throw like `GetPageAsync` or `GetLeafAsycn`
2. Rename `GetIndexAsync` to something like `GetIndexOrNullAsync`.

File: src/NuGet.Services.AzureSearch/Registration/Models/RegistrationLeaf.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public class RegistrationLeaf
+    {
+        [JsonProperty("@id")]
+        public string Url { get; set; }
+
+        [JsonProperty("catalogEntry")]
+        public string CatalogEntry { get; set; }
+
+        [JsonProperty("list")]
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/Registration/Models/RegistrationLeaf.cs]: >list [](start = 23, length = 4)

`list` -> `listed`

**Comment by loic-sharma** [comment]
>     <value>{"@id":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/index.json","@type":["catalog:CatalogRoot","PackageRegistration","catalog:Permalink"],"commitId":"bef2767e-f5ae-4713-b5fc-510945bacdf9","commitTimeStamp":"2018-11-26T19:53:07.9393299Z","count":1,"items":[{"@id":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/index.json#page/0.1.1/0.3.1","@type":"catalog:CatalogPage","commitId":"bef2767e-f5ae-4713-b5fc-510945bacdf9","commitTimeStamp":"2018-11-26T19:53:07.9393299Z","count":4,"items":[{"@id":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/0.1.1.json","@type":"Package","commitId":"bef2767e-f5ae-4713-b5fc-510945bacdf9","commitTimeStamp":"2018-11-26T19:53:07.9393299Z","catalogEntry":{"@id":"https://api.nuget.org/v3/catalog0/data/2018.11.13.04.43.04/microbuild.core.0.1.1.json","@type":"PackageDetails","authors":"Microsoft","description":"MicroBuild bootstrapper package which wires up build targets that load and execute other MicroBuild plugins during the build","iconUrl":"","id":"MicroBuild.Core","language":"","licenseUrl":"","listed":false,"minClientVersion":"","packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.1.1/microbuild.core.0.1.1.nupkg","projectUrl":"","published":"1900-01-01T00:00:00+00:00","requireLicenseAcceptance":false,"summary":"","tags":["MicroBuild"],"title":"","version":"0.1.1"},"packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.1.1/microbuild.core.0.1.1.nupkg","registration":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/index.json"},{"@id":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/0.2.0.json","@type":"Package","commitId":"bef2767e-f5ae-4713-b5fc-510945bacdf9","commitTimeStamp":"2018-11-26T19:53:07.9393299Z","catalogEntry":{"@id":"https://api.nuget.org/v3/catalog0/data/2018.11.13.04.43.04/microbuild.core.0.2.0.json","@type":"PackageDetails","authors":"Microsoft","description":"MicroBuild bootstrapper package which wires up build targets that load and execute other MicroBuild plugins during the build","iconUrl":"","id":"MicroBuild.Core","language":"","licenseUrl":"http://microsoft.mit-license.org/","listed":false,"minClientVersion":"","packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.2.0/microbuild.core.0.2.0.nupkg","projectUrl":"","published":"1900-01-01T00:00:00+00:00","requireLicenseAcceptance":false,"summary":"","tags":["MicroBuild"],"title":"","version":"0.2.0"},"packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.2.0/microbuild.core.0.2.0.nupkg","registration":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/index.json"},{"@id":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/0.3.0.json","@type":"Package","commitId":"bef2767e-f5ae-4713-b5fc-510945bacdf9","commitTimeStamp":"2018-11-26T19:53:07.9393299Z","catalogEntry":{"@id":"https://api.nuget.org/v3/catalog0/data/2018.11.26.19.52.28/microbuild.core.0.3.0.json","@type":"PackageDetails","authors":"Microsoft","description":"MicroBuild bootstrapper package which wires up build targets that load and execute other MicroBuild plugins during the build","iconUrl":"","id":"MicroBuild.Core","language":"","licenseUrl":"http://microsoft.mit-license.org/","listed":false,"minClientVersion":"","packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.3.0/microbuild.core.0.3.0.nupkg","projectUrl":"","published":"1900-01-01T00:00:00+00:00","requireLicenseAcceptance":false,"summary":"","tags":["MicroBuild"],"title":"","version":"0.3.0"},"packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.3.0/microbuild.core.0.3.0.nupkg","registration":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/index.json"},{"@id":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/0.3.1.json","@type":"Package","commitId":"bef2767e-f5ae-4713-b5fc-510945bacdf9","commitTimeStamp":"2018-11-26T19:53:07.9393299Z","catalogEntry":{"@id":"https://api.nuget.org/v3/catalog0/data/2018.11.26.19.49.07/microbuild.core.0.3.1.json","@type":"PackageDetails","authors":"Microsoft","description":"MicroBuild bootstrapper package which wires up build targets that load and execute other MicroBuild plugins during the build","iconUrl":"https://github.com/MicrosoftDocs/Microbuild.Core/blob/master/extension-icon.png?raw=true","id":"MicroBuild.Core","language":"en-US","licenseUrl":"http://microsoft.mit-license.org/","listed":true,"minClientVersion":"","packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.3.1/microbuild.core.0.3.1.nupkg","projectUrl":"https://aka.ms/microbuild.package","published":"2018-11-13T04:26:56.943+00:00","requireLicenseAcceptance":true,"summary":"","tags":["MicroBuild"],"title":"","version":"0.3.1"},"packageContent":"https://api.nuget.org/v3-flatcontainer/microbuild.core/0.3.1/microbuild.core.0.3.1.nupkg","registration":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/index.json"}],"parent":"https://api.nuget.org/v3/registration3-gz-semver2/microbuild.core/index.json","lower":"0.1.1","upper":"0.3.1"}],"@context":{"@vocab":"http://schema.nuget.org/schema#","catalog":"http://schema.nuget.org/catalog#","xsd":"http://www.w3.org/2001/XMLSchema#","items":{"@id":"catalog:item","@container":"@set"},"commitTimeStamp":{"@id":"catalog:commitTimeStamp","@type":"xsd:dateTime"},"commitId":{"@id":"catalog:commitId"},"count":{"@id":"catalog:count"},"parent":{"@id":"catalog:parent","@type":"@id"},"tags":{"@container":"@set","@id":"tag"},"packageTargetFrameworks":{"@container":"@set","@id":"packageTargetFramework"},"dependencyGroups":{"@container":"@set","@id":"dependencyGroup"},"dependencies":{"@container":"@set","@id":"dependency"},"packageContent":{"@type":"@id"},"published":{"@type":"xsd:dateTime"},"registration":{"@type":"@id"}}}</value>

Consider making `listed` have a value of `true` so that you can test it was actually deserialized

---
Refers to: tests/NuGet.Protocol.Catalog.Tests/TestData.resx:294 in f2120a6. [](commit_id = f2120a61492b6da0afad930ecc6b842fbf578b2d, deletion_comment = False)

**Comment by joelverhagen** [comment]
> Consider making `listed` have a value of true so that you can test it was actually deserialized

Good idea.

File: src/NuGet.Services.AzureSearch/Registration/IRegistrationClient.cs (hunk 1 above)
Comment (loic-sharma) [reply src/NuGet.Services.AzureSearch/Registration/IRegistrationClient.cs]: Oh right I forgot about that client package. Yeah I guess there's no need for this right now, ignore my comment.

**Pull request merged**